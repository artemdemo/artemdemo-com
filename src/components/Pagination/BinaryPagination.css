@media (max-width: 768px) {
    .binary-pagination {
        display: block;
    }
}

.binary-pagination-link {
    max-width: 200px;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
}

@media (min-width: 992px) {
    .binary-pagination-link {
        max-width: 300px;
    }
}