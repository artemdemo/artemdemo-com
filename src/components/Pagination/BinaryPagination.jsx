import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'gatsby';

class BinaryPagination extends React.PureComponent {
    renderPrev() {
        const { previous } = this.props;
        if (previous && previous.slug != null) {
            const title = previous.title || previous.slug;
            return (
                <li>
                    <Link to={previous.slug} rel="prev">
                        ← {title}
                    </Link>
                </li>
            );
        }
        return null;
    }

    renderNext() {
        const { next } = this.props;
        if (next && next.slug != null) {
            const title = next.title || next.slug;
            return (
                <li>
                    <Link to={next.slug} rel="next">
                        {title} →
                    </Link>
                </li>
            );
        }
        return null;
    }

    render() {
        return (
            <ul
                style={{
                    display: 'flex',
                    flexWrap: 'wrap',
                    justifyContent: 'space-between',
                    listStyle: 'none',
                    padding: 0,
                }}
            >
                {this.renderPrev()}
                {this.renderNext()}
            </ul>
        );
    }
}

BinaryPagination.propTypes = {
    previous: PropTypes.shape({
        slug: PropTypes.string,
        title: PropTypes.string,
    }),
    next: PropTypes.shape({
        slug: PropTypes.string,
        title: PropTypes.string,
    }),
};

BinaryPagination.defaultProps = {
    previous: null,
    next: null,
};

export default BinaryPagination;