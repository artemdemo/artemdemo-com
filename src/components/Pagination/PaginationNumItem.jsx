import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'gatsby';
import * as utils from '../../services/utils';

const PaginationNumItem = (props) => {
    const { pageNum, active } = props;
    const itemClass = classnames('page-item', {
        active
    });
    const path = pageNum > 1 ? utils.prefixBlog(`/${pageNum}`) : '/';

    return (
        <li className={itemClass}>
            <Link
                className='page-link'
                to={path}
            >
                {pageNum}
            </Link>
        </li>
    );
};

PaginationNumItem.propTypes = {
    pageNum: PropTypes.number.isRequired,
    active: PropTypes.bool,
};

PaginationNumItem.defaultProps = {
    active: false,
};

export default PaginationNumItem;