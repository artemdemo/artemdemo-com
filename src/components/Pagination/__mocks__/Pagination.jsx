import React from 'react';

const Pagination = props => (
    <div data-mock='Pagination'>
        {JSON.stringify(props, null, 2)}
    </div>
);

export default Pagination;