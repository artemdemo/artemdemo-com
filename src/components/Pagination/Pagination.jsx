import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Link } from 'gatsby';
import * as utils from '../../services/utils';
import PaginationNumItem from './PaginationNumItem';

class Pagination extends React.PureComponent {
    prefixPath(path) {
        const { prefixer } = this.props;
        return prefixer ? prefixer(path) : path;
    }

    renderPrev() {
        const { skip, totalPosts, limit } = this.props;
        if (totalPosts === 0) {
            return null;
        }
        const itemClass = classnames('page-item', {
            disabled: skip === 0
        });
        const prevPageNumber = Math.floor(skip / limit);
        const prevPath = prevPageNumber > 1 ? this.prefixPath(`/${prevPageNumber}`) : this.prefixPath();
        return (
            <li className={itemClass}>
                <Link
                    className='page-link'
                    to={prevPath}
                >
                    Previous
                </Link>
            </li>
        );
    }

    renderNext() {
        const { skip, totalPosts, limit} = this.props;
        if (totalPosts === 0) {
            return null;
        }
        const pagesAmount = Math.ceil(totalPosts / limit);
        const nextPageNumber = Math.floor(skip / limit) + 2;
        const itemClass = classnames('page-item', {
            disabled: nextPageNumber > pagesAmount,
        });
        return (
            <li className={itemClass}>
                <Link
                    className='page-link'
                    to={this.prefixPath(`/${nextPageNumber}`)}
                >
                    Next
                </Link>
            </li>
        );
    }

    renderNumPages() {
        const { totalPosts, limit, skip, prefixer } = this.props;
        if (totalPosts === 0) {
            return null;
        }
        const pagesAmount = Math.ceil(totalPosts / limit);
        const currentNumber = Math.floor(skip / limit) + 1;
        const iterArr = Array.from(Array(pagesAmount));
        return (
            <React.Fragment>
                {iterArr.map((_, index) => (
                    <PaginationNumItem
                        pageNum={index + 1}
                        active={currentNumber === index + 1}
                        prefixer={prefixer}
                        key={`pagination-item-${index}`}
                    />
                ))}
            </React.Fragment>
        );
    }

    render() {
        const { totalPosts } = this.props;

        if (totalPosts <= utils.POSTS_PER_PAGE) {
            return null;
        }

        return (
            <ul className='pagination justify-content-center'>
                {this.renderPrev()}
                {this.renderNumPages()}
                {this.renderNext()}
            </ul>
        );
    }
}

Pagination.propTypes = {
    skip: PropTypes.number,
    limit: PropTypes.number,
    totalPosts: PropTypes.number,
    prefixer: PropTypes.func,
};

Pagination.defaultProps = {
    skip: 0,
    limit: 0,
    totalPosts: 0,
    prefixer: null,
};

export default Pagination;