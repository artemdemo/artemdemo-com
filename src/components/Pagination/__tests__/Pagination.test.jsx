import React from 'react';
import renderer from 'react-test-renderer';
import Pagination from '../Pagination';

jest.mock('gatsby');

describe('Pagination', () => {
    it('should render empty', () => {
        const tree = renderer.create(
            <Pagination />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with data', () => {
        const tree = renderer.create(
            <Pagination
                previous={{
                    slug: '/prev',
                    title: 'Prev title',
                }}
                next={{
                    slug: '/next',
                    title: 'Next title',
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});