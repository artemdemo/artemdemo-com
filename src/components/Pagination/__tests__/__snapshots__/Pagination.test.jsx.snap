// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination default render (empty) 1`] = `null`;

exports[`Pagination should render 1 page out of 2 1`] = `
<ul
  className="pagination justify-content-center"
>
  <li
    className="page-item disabled"
  >
    <div
      data-mock="Link"
    >
      {
  "className": "page-link"
}
      Previous
    </div>
  </li>
  <div
    data-mock="PaginationNumItem"
  >
    {
  "pageNum": 1,
  "active": true,
  "prefixer": null
}
  </div>
  <div
    data-mock="PaginationNumItem"
  >
    {
  "pageNum": 2,
  "active": false,
  "prefixer": null
}
  </div>
  <li
    className="page-item"
  >
    <div
      data-mock="Link"
    >
      {
  "className": "page-link",
  "to": "/2"
}
      Next
    </div>
  </li>
</ul>
`;

exports[`Pagination should render 2 page out of 2 1`] = `
<ul
  className="pagination justify-content-center"
>
  <li
    className="page-item"
  >
    <div
      data-mock="Link"
    >
      {
  "className": "page-link"
}
      Previous
    </div>
  </li>
  <div
    data-mock="PaginationNumItem"
  >
    {
  "pageNum": 1,
  "active": false,
  "prefixer": null
}
  </div>
  <div
    data-mock="PaginationNumItem"
  >
    {
  "pageNum": 2,
  "active": true,
  "prefixer": null
}
  </div>
  <li
    className="page-item disabled"
  >
    <div
      data-mock="Link"
    >
      {
  "className": "page-link",
  "to": "/3"
}
      Next
    </div>
  </li>
</ul>
`;
